cts/cts-regression.in

```diff
@@ -1,230 +1,257 @@
+        except KeyError:
+            pass
+
+        setattr(namespace, self.dest, sorted(list(tests)))
+
+
+def error_print(msg):
+    """Print an error message.
+
+    :param msg: Message to print
+    :type msg: str
+    """
+    print('      * ERROR:   %s' % msg)
+
+
+def run_test(test, test_home, verbose=False, valgrind=False):
+    """Run a regression test and return the result.
+
+    :param test: The name of the test to run
+    :type test: str
+    :return: The exit code of the test
+    :rtype: :class:`CrmExit`
+    """
+    print('Executing the %s regression tests' % test)
+    print('=' * 60)
+    if test == 'pacemaker_remote':
+        cmd = [os.path.join(test_home, 'cts-exec'), '-R']
+    else:
+        cmd = [os.path.join(test_home, 'cts-%s' % test)]
+
+    if verbose:
+        cmd.append('--verbose')
+
+    # Valgrind is supported only for cli and scheduler currently
+    if valgrind and test in ['cli', 'scheduler']:
+        cmd.append('--valgrind')
+
+    if os.access(cmd[0], os.X_OK):
+        # Test might not be executable if run from source directory
+        mode = os.stat(cmd[0]).st_mode
+        os.chmod(cmd[0], mode | stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)
+
+        # Run as root
+        if os.geteuid() != 0:
+            print('Enter the sudo password if prompted')
+            cmd = ['sudo'] + cmd
+
+        rc = CrmExit(subprocess.call(cmd))
+    else:
+        error_print('%s regression test not found or not executable' % test)
+        rc = CrmExit.NOT_INSTALLED
+
+    print('=' * 60 + '\n\n')
+    return rc
+
+
+def run_tests(tests, test_home, verbose=False, valgrind=False):
+    """Run regression tests and report results for each.
+
+    :param tests: A list of names of tests to run
+    :type tests: list[str]
+    :return: :attr:`CrmExit.OK` if all tests were successful,
+        :attr:`CrmExit.ERROR` otherwise
+    :rtype: :class:`CrmExit`
+    """
+    failed = []
+
+    for test in tests:
+        rc = run_test(test, test_home, verbose, valgrind)
+        if rc != CrmExit.OK:
+            error_print('%s regression tests failed (%s)' % (test, rc))
+            failed.append(test)
+
+    if failed:
+        print('Failed regression tests:', end='')
+        for test in failed:
+            print(' %s' % test, end='')
+        print()
+        return CrmExit.ERROR
+
+    return CrmExit.OK
+
+
+def main():
+    """Run Pacemaker regression tests as specified by arguments."""
+    description = textwrap.dedent('''\
+        Run Pacemaker regression tests.
+
+        Available tests (default tests are 'scheduler cli'):
+         scheduler         Action scheduler
+         cli               Command-line tools
+         exec              Local resource agent executor
+         fencing           Fencer
+         all               Synonym for 'scheduler cli exec fencing'
+    ''')
+    if REMOTE_ENABLED:
+        description += (
+            ' pacemaker_remote  Resource agent executor in remote mode'
```

Comment:
This can be moved underneath `exec` if we really want to, with `fencing` and `all` underneath `pacemaker_remote`. This reordering made the string formatting a lot cleaner and I figure it doesn't matter much.

Reply:
The description can be a component class member as well, and this reduces nicely

Reply:
Done.